test/integration/element_tracing.cc

```diff
@@ -0,0 +1,291 @@
+  EXPECT_EQ(overrideActorPluginXmlPath, overrideActorPluginElem->XmlPath());
+
+  // Lights
+  const std::string lightFilePath = sdf::testing::TestFile(
+      "integration", "model", "test_light", "model.sdf");
+  const std::string lightXmlPath = "/sdf/light[@name=\"point_light\"]";
+
+  const sdf::Light *light = world->LightByIndex(0);
+  ASSERT_NE(nullptr, light);
+  sdf::ElementPtr lightElem = light->Element();
+  ASSERT_NE(nullptr, lightElem);
+  EXPECT_EQ(lightFilePath, lightElem->FilePath());
+  ASSERT_TRUE(lightElem->LineNumber().has_value());
+  EXPECT_EQ(3, lightElem->LineNumber().value());
+  EXPECT_EQ(lightXmlPath, lightElem->XmlPath());
+
+  const std::string overrideLightXmlPath =
+      "/sdf/light[@name=\"override_light_name\"]";
+  const sdf::Light *overrideLight = world->LightByIndex(1);
+  ASSERT_NE(nullptr, overrideLight);
+  sdf::ElementPtr overrideLightElem = overrideLight->Element();
+  ASSERT_NE(nullptr, overrideLightElem);
+  EXPECT_EQ(lightFilePath, overrideLightElem->FilePath());
+  ASSERT_TRUE(overrideLightElem->LineNumber().has_value());
+  EXPECT_EQ(3, overrideLightElem->LineNumber().value());
+  EXPECT_EQ(overrideLightXmlPath, overrideLightElem->XmlPath());
+
+  // Models
+  const std::string modelFilePath = sdf::testing::TestFile(
+      "integration", "model", "test_model", "model.sdf");
+  const std::string modelXmlPath = "/sdf/model[@name=\"test_model\"]";
+
+  const sdf::Model *model = world->ModelByIndex(0);
+  ASSERT_NE(nullptr, model);
+  sdf::ElementPtr modelElem = model->Element();
+  ASSERT_NE(nullptr, modelElem);
+  EXPECT_EQ(modelFilePath, modelElem->FilePath());
+  ASSERT_TRUE(modelElem->LineNumber().has_value());
+  EXPECT_EQ(4, modelElem->LineNumber().value());
+  EXPECT_EQ(modelXmlPath, modelElem->XmlPath());
+
+  const std::string overrideModelXmlPath =
+      "/sdf/model[@name=\"override_model_name\"]";
+  const sdf::Model *overrideModel = world->ModelByIndex(1);
+  ASSERT_NE(nullptr, overrideModel);
+  sdf::ElementPtr overrideModelElem = overrideModel->Element();
+  ASSERT_NE(nullptr, overrideModelElem);
+  EXPECT_EQ(modelFilePath, overrideModelElem->FilePath());
+  ASSERT_TRUE(overrideModelElem->LineNumber().has_value());
+  EXPECT_EQ(4, overrideModelElem->LineNumber().value());
+  EXPECT_EQ(overrideModelXmlPath, overrideModelElem->XmlPath());
+
+  const std::string overrideModelPluginXmlPath =
+      worldXmlPath + "/include[1]/plugin[0]";
+
+  sdf::ElementPtr overrideModelPluginElem =
+      overrideModelElem->GetElement("plugin");
+  ASSERT_NE(nullptr, overrideModelPluginElem);
+  EXPECT_EQ(worldFile, overrideModelPluginElem->FilePath());
+  ASSERT_TRUE(overrideModelPluginElem->LineNumber().has_value());
+  EXPECT_EQ(14, overrideModelPluginElem->LineNumber().value());
+  EXPECT_EQ(overrideModelPluginXmlPath, overrideModelPluginElem->XmlPath());
+
+  const std::string modelWithFileFilePath = sdf::testing::TestFile(
+      "integration", "model", "test_model/model.sdf");
```

Comment:
Ah, this was the cause of one of the failures in CI, due to the behavior of these lines, https://github.com/osrf/sdformat/blob/sdf11/test/test_config.h.in#L102-L122. It appends the model name, which in Windows, will turn into `Some\\Path\\To\\Model\\test_model/model.sdf`, due to the name overload as a path, https://github.com/osrf/sdformat/blob/sdf11/test/sdf/includes.sdf#L18

https://build.osrfoundation.org/job/sdformat-ci-pr_any-windows7-amd64/2181/consoleFull#45593342ed30c675-ba23-4c35-b655-f5c948f97581

I would say this only applies to this particular test case.

Reply:
I'm a little confused about this. I agree with Jenn's suggestion that we shouldn't use `/` in `sdf::testing::TestFile`. Are you saying even when **not** using `/` in `sdf::testing::TestFile`, `sdf::Element::FilePath()` returns `Some\\Path\\To\\Model\\test_model/model.sdf` (with a `/`)?

Reply:
Yeah! It doesn't split `test_model/model.sdf` into `test_model` and `model.sdf`, but just appends it, therefore even in the Windows7 build, we get the path with `\\path\\test_model/model.sdf` phenomenon.

```
17: [ RUN      ] ElementTracing.includes
17: C:\Jenkins\workspace\sdformat-ci-pr_any-windows7-amd64\ws\sdformat\test\integration\element_tracing.cc(283): error: Expected equality of these values:
17:   modelFilePath
17:     Which is: "C:/Jenkins/workspace/sdformat-ci-pr_any-windows7-amd64/ws/sdformat\\test\\integration\\model\\test_model\\model.sdf"
17:   overrideModelWithFileElem->FilePath()
17:     Which is: "C:/Jenkins/workspace/sdformat-ci-pr_any-windows7-amd64/ws/sdformat\\test\\integration\\model\\test_model/model.sdf"
17: [  FAILED  ] ElementTracing.includes (201 ms)
```

Reply:
I see. `sdf::findFile` should probably sanitize the file path when translating URIs (which should always use `/`) into filepaths. Do you mind (1) adding a comment to the line 276 in the test explaining why there's a `/` and (2) creating an issue stating that filepaths returned by `sdf::fineFile` should be sanitized?

Reply:
@jennuine @azeey, after digging a little deeper, I found a work-around by adding some minor path sanitization in `findFileCb` that was defined in `element_tracing.cc`, https://github.com/osrf/sdformat/pull/548/commits/bd721e99ed2c388e6bd2545059e469115e965aae. This will at least allow us to not use `/` in `sdf::testing::TestFile`.

I will be creating the issue soon, thanks!

Reply:
@azeey The outputs of `sdf::findFile` is sanitized [here](https://github.com/osrf/sdformat/blob/sdf11/src/SDF.cc#L60-L86), at this point, it is just `sdf::testing::TestFile` that needs to handle sanitization, https://github.com/osrf/sdformat/issues/572

Reply:
I posted on #572, but the sanitization you linked to only applies when the URI is in `URIPathMap`. It doesn't do that for regular file paths.